pkg/controller/utils/component.go

```diff
@@ -85,6 +92,111 @@ func (c *componentHandler) SetCreateOnly() {
 	c.createOnly = true
 }
 
+func (c *componentHandler) create(ctx context.Context, obj client.Object, opts ...client.CreateOption) error {
+	// Make a deep copy of the object, so we can stash away the original object in the cache.
+	cp := obj.DeepCopyObject().(client.Object)
+
+	// Pass to the client.
+	err := c.client.Create(ctx, obj, opts...)
+	if err != nil {
+		return err
+	}
+
+	// Update the caches so that we don't try to update the object on subsequent reconciliations.
+	dCache.set(cp, obj.GetGeneration())
+	return nil
+}
+
+func (c *componentHandler) update(ctx context.Context, obj client.Object, opts ...client.UpdateOption) error {
+	logCtx := ContextLoggerForResource(c.log, obj)
+	if !c.needsUpdate(ctx, obj) {
+		// The object does not need to be updated, so we can skip it.
+		logCtx.V(2).Info("Object does not need to be updated, skipping")
+		return nil
+	}
+	logCtx.V(2).Info("Object needs to be updated")
+
+	// Make a deep copy of the object, so we can stash away the original object in the cache.
+	cp := obj.DeepCopyObject().(client.Object)
+
+	// Pass to the client.
+	err := c.client.Update(ctx, obj, opts...)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Inlalidate our cached object if it was not found.
+			dCache.delete(obj)
+		}
+		return err
+	}
+
+	// Update the caches so that we don't try to update the object on subsequent reconciliations.
+	dCache.set(cp, obj.GetGeneration())
+	return nil
+}
+
+func (c *componentHandler) delete(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error {
+	err := c.client.Delete(ctx, obj, opts...)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Remove the object from the cache if it was not found.
+			dCache.delete(obj)
+		}
+		return err
+	}
+
+	// Invalidate our cached object.
+	dCache.delete(obj)
+	return nil
+}
+
+func (c *componentHandler) needsUpdate(ctx context.Context, obj client.Object) bool {
+	key := client.ObjectKeyFromObject(obj)
+	cur := obj.DeepCopyObject().(client.Object)
+	err := c.client.Get(ctx, key, cur)
```

Comment:
Wondering if you considered removing this `Get`? From my understanding, when we reach this point in the code, we already know that the object exists, otherwise we would have hit the create branch in `createOrUpdateObject`. We could also utilize the generation from the obj passed in. Perhaps you've implemented it this way to ensure we have the freshest information on the object?

Reply:
I hadn't considered it - I think it should be possible? It maybe widens the possibility of using out-of-date data, but shouldn't by much. 

That said, having the get here should be more or less free since the client is served from a cache.

I think we'd just need to ensure the generation is set as part of `mergeState` and then this function could rely on that field. Would clean things up here a fair bit!